File: backend/api/schemas/transaction.py
from pydantic import BaseModel, model_validator

from api.core.pii_masker import mask_account_number


class TransactionBase(BaseModel):
    """Shared transaction fields."""

    transaction_date: str
    transaction_type: str
    amount: float
    currency: str = "INR"
    direction: str
    channel: str | None = None
    counterparty_name: str | None = None
    counterparty_account: str | None = None
    counterparty_bank: str | None = None
    location: str | None = None
    country: str | None = None
    reference_number: str | None = None
    description: str | None = None
    is_flagged: bool = False


class TransactionResponse(TransactionBase):
    """Transaction data returned from API."""

    id: str
    account_id: str

    model_config = {"from_attributes": True}


class MaskedTransactionResponse(TransactionResponse):
    """Transaction response with counterparty_account masked for DPDP Act compliance."""

    @model_validator(mode="after")
    def apply_pii_masking(self) -> "MaskedTransactionResponse":
        """Mask the counterparty account number."""
        self.counterparty_account = mask_account_number(self.counterparty_account)
        return self

File: frontend/src/types/transaction.ts
export interface Transaction {
  id: string;
  account_id: string;
  transaction_date: string;
  transaction_type: string;
  amount: number;
  currency: string;
  direction: string;
  channel: string | null;
  counterparty_name: string | null;
  counterparty_account: string | null;
  counterparty_bank: string | null;
  location: string | null;
  country: string | null;
  reference_number: string | null;
  description: string | null;
  is_flagged: boolean;
}
